### ISSUE TITLE
Cannot reuse Workflow ID immediately

### ISSUE THREAD

**Issue by si-stripe** [temporalio/sdk-go#492]
## Expected Behavior
With a `WORKFLOW_ID_REUSE_POLICY_ALLOW_DUPLICATE` reuse policy, a Workflow ID should be able to be reused as soon as the Workflow before it completes.

## Actual Behavior
If I immediately try and reuse a Workflow ID I get:

```
2021/07/19 23:53:58 ERROR Workflow panic Namespace default TaskQueue ---- WorkerID --- WorkflowType WorkflowManagerWorkflow WorkflowID someWorkflowId-manager RunID c68cd12c-3761-4e8b-9b9c-44e541216a8f Attempt 1 Error adding duplicate command CommandType: ChildWorkflow, ID: someWorkflowId, state=Created, isDone()=false, history=[Created] StackTrace coroutine root [panic]:
go.temporal.io/sdk/internal.panicIllegalState(...)
        external/io_temporal_go_sdk/internal/internal_decision_state_machine.go:393
go.temporal.io/sdk/internal.(*commandsHelper).addCommand(0xc00013f400, 0x1de05c8, 0xc000635780)
        external/io_temporal_go_sdk/internal/internal_decision_state_machine.go:841 +0x349
go.temporal.io/sdk/internal.(*commandsHelper).startChildWorkflowExecution(0xc00013f400, 0xc000646500, 0x0, 0x0)
        external/io_temporal_go_sdk/internal/internal_decision_state_machine.go:1021 +0x59
go.temporal.io/sdk/internal.(*workflowEnvironmentImpl).ExecuteChildWorkflow(0xc000f53680, 0xc000ec8280, 0x8, 0x0, 0x0, 0x0, 0xc000d66728, 0x7, 0xc00063e7e0, 0xe, ...)
        external/io_temporal_go_sdk/internal/internal_event_handlers.go:395 +0x2e9
go.temporal.io/sdk/internal.(*workflowEnvironmentInterceptor).ExecuteChildWorkflow(0xc0001434a0, 0x1dd2ee8, 0xc000fb0ed0, 0x1fcee8d, 0x10, 0xc000635680, 0x1, 0x1, 0x10, 0x11b0240)
        external/io_temporal_go_sdk/internal/workflow.go:708 +0x86d
go.temporal.io/sdk/internal.ExecuteChildWorkflow(0x1dd2ee8, 0xc000fb0ed0, 0x11e6740, 0x1b86f88, 0xc000635680, 0x1, 0x1, 0x0, 0x0)
        external/io_temporal_go_sdk/internal/workflow.go:671 +0x193
go.temporal.io/sdk/workflow.ExecuteChildWorkflow(...)
        external/io_temporal_go_sdk/workflow/workflow.go:182
reflect.Value.call(0x11bab60, 0x1b86f98, 0x13, 0x135e73d, 0x4, 0xc000f0edf0, 0x1, 0x1, 0x1d97740, 0x133c320, ...)
        GOROOT/src/reflect/value.go:476 +0x8e7
reflect.Value.Call(0x11bab60, 0x1b86f98, 0x13, 0xc000f0edf0, 0x1, 0x1, 0x1d97740, 0x133c320, 0xc0001434a0)
        GOROOT/src/reflect/value.go:337 +0xb9
go.temporal.io/sdk/internal.(*workflowEnvironmentInterceptor).ExecuteWorkflow(0xc0001434a0, 0x1dd2d60, 0xc00013f540, 0xc000984108, 0x17, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
        external/io_temporal_go_sdk/internal/workflow.go:391 +0x2cb
go.temporal.io/sdk/internal.(*workflowExecutor).Execute(0xc00013f4c0, 0x1dd2d60, 0xc00013f540, 0x0, 0xc000dcc738, 0xc8c8e5, 0x0)
        external/io_temporal_go_sdk/internal/internal_worker.go:755 +0x316
go.temporal.io/sdk/internal.(*syncWorkflowDefinition).Execute.func1(0x1dd2ee8, 0xc000f4fad0)
        external/io_temporal_go_sdk/internal/internal_workflow.go:494 +0xf5
2021/07/19 23:53:58 WARN  Failed to process workflow task. Namespace default TaskQueue ---- WorkerID --- WorkflowType WorkflowManagerWorkflow WorkflowID someWorkflowId-manager RunID c68cd12c-3761-4e8b-9b9c-44e541216a8f Attempt 1 Error adding duplicate command CommandType: ChildWorkflow, ID: someWorkflowId, state=Created, isDone()=false, history=[Created]
```

## Steps to Reproduce the Problem

  1. Create a Workflow that starts a child workflow with ID `ChildWorkflow`
  2. Have ChildWorkflow return and send a signal to Parent workflow that it's done
  3. When Parent workflow receives done signal, start a new child workflow with ID `ChildWorkflow`

## Specifications

  - Version: 1.9.2
  - Platform: Go

**Comment by mfateev** [comment]
In your scenario, the child is not completed yet when the signal is received. So the attempt to schedule another child with the same ID fails per design.

Waiting on the child's future is the supported way to learn about a child's completion. I don't understand why you are trying to use a signal for this.

**Comment by si-stripe** [comment]
I wanted to use signals/select to control my Workflow without polling. 

I can achieve this by creating a goroutine that blocks on `future.Get()` and then sends a signal to indicate that the child is done. If you think it's a common pattern, I'm happy to add some docs for this

**Comment by Drahflow** [comment]
We hit the same issue today, and I became convinced the issue has been closed in error. The internal state machine command IDs for child workflows depend only on the child's workflow ID:
https://github.com/temporalio/sdk-go/blob/a23dec9aba94a515920e3d298ce29190316b295d/internal/internal_command_state_machine.go#L429,L435
... but if the same workflow ever tries to create two children with the same ID (which is legal if it made sure that the first has completed before) it will still panic.

**Comment by cretz** [comment]
> The internal state machine command IDs for child workflows depend only on the child's workflow ID

It makes the command based on child workflow ID and an internal counter it maintains

> but if the same workflow ever tries to create two children with the same ID (which is legal if it made sure that the first has completed before) it will still panic.

Can you provide a small standalone replication of this? Ideally in a new issue.

**Comment by cretz** [comment]
@Drahflow - Reopening as I believe this issue probably does exist. In the meantime as a workaround can you use unique child workflow IDs for each child workflow invocation?

**Comment by cretz** [comment]
@Drahflow - Can you provide a small standalone replication? Can you make sure you are waiting for child workflow completion in this replication?

**Comment by cretz** [comment]
Ok, after attempted replication, I am able to create children with the same ID. Please make sure you confirm the previous one is complete. Here's an example of creating a child every 5 seconds:

```go
package main

import (
	"context"
	"fmt"
	"log"
	"time"

	"github.com/google/uuid"
	"go.temporal.io/sdk/client"
	"go.temporal.io/sdk/worker"
	"go.temporal.io/sdk/workflow"
)

func main() {
	if err := run(); err != nil {
		log.Fatal(err)
	}
}

func run() error {
	ctx, cancel := context.WithCancel(context.Background())
	defer cancel()

	log.Printf("Creating client")
	c, err := client.NewClient(client.Options{})
	if err != nil {
		return fmt.Errorf("failed creating client: %w", err)
	}
	defer c.Close()

	log.Printf("Starting worker")
	taskQueue := uuid.NewString()
	worker := worker.New(c, taskQueue, worker.Options{})
	worker.RegisterWorkflow(MyWorkflow)
	worker.RegisterWorkflow(MyChildWorkflow)
	if err := worker.Start(); err != nil {
		return fmt.Errorf("failed starting worker client: %w", err)
	}

	log.Printf("Starting workflow")
	run, err := c.ExecuteWorkflow(ctx, client.StartWorkflowOptions{TaskQueue: taskQueue}, MyWorkflow)
	if err != nil {
		return fmt.Errorf("failed starting workflow: %w", err)
	}

	return run.Get(ctx, nil)
}

func MyWorkflow(ctx workflow.Context) error {
	// Start children with the same ID over and over
	ctx = workflow.WithChildOptions(ctx, workflow.ChildWorkflowOptions{
		WorkflowID: workflow.GetInfo(ctx).WorkflowExecution.ID + "-child",
	})
	for {
		workflow.GetLogger(ctx).Info("Starting child and waiting for completion")
		if err := workflow.ExecuteChildWorkflow(ctx, MyChildWorkflow).Get(ctx, nil); err != nil {
			return err
		}
	}
}

func MyChildWorkflow(ctx workflow.Context) error {
	return workflow.Sleep(ctx, 5*time.Second)
}
```
